File: test/data/zig/sample.zig
//! Top-level doc comment for the module
const std = @import("std");
const math = std.math;

/// A simple Point struct
pub const Point = struct {
    x: f32,
    y: f32,
    
    /// Calculate distance from origin
    pub fn distanceFromOrigin(self: Point) f32 {
        return math.sqrt(self.x * self.x + self.y * self.y);
    }
    
    pub fn add(self: Point, other: Point) Point {
        return Point{
            .x = self.x + other.x,
            .y = self.y + other.y,
        };
    }
};

/// Color enum with RGB values
pub const Color = enum {
    red,
    green,
    blue,
    
    pub fn toRGB(self: Color) u32 {
        return switch (self) {
            .red => 0xFF0000,
            .green => 0x00FF00,
            .blue => 0x0000FF,
        };
    }
};

/// Tagged union for shapes
pub const Shape = union(enum) {
    circle: f32,
    rectangle: struct { width: f32, height: f32 },
    
    pub fn area(self: Shape) f32 {
        return switch (self) {
            .circle => |radius| math.pi * radius * radius,
            .rectangle => |rect| rect.width * rect.height,
        };
    }
};

/// Error set for file operations
pub const FileError = error{
    NotFound,
    AccessDenied,
    InvalidPath,
};

// Global constant
const MAX_SIZE: usize = 1024;

// Mutable variable
var global_counter: u32 = 0;

/// Main entry point
pub fn main() !void {
    const allocator = std.heap.page_allocator;
    
    // Variable declarations
    var x: i32 = 42;
    const y: i32 = 100;
    
    // Control flow
    if (x > 0) {
        x += 1;
    } else {
        x = 0;
    }
    
    // Loop
    var i: u32 = 0;
    while (i < 10) : (i += 1) {
        if (i == 5) continue;
        if (i == 8) break;
    }
    
    // For loop
    const items = [_]i32{ 1, 2, 3, 4, 5 };
    for (items) |item| {
        _ = item;
    }
    
    // Error handling with try
    const file = try std.fs.cwd().openFile("test.txt", .{});
    defer file.close();
    
    // Builtin functions
    const size = @sizeOf(Point);
    _ = @intCast(i64, x);
    
    // Comptime
    comptime {
        const compile_time_value = 42 * 2;
        _ = compile_time_value;
    }
    
    _ = allocator;
    _ = y;
    _ = size;
}

/// Generic function
pub fn swap(comptime T: type, a: *T, b: *T) void {
    const temp = a.*;
    a.* = b.*;
    b.* = temp;
}

/// Async function example
pub fn asyncOperation() callconv(.Async) void {
    suspend {}
}

test "basic test" {
    const result = 2 + 2;
    try std.testing.expect(result == 4);
}

test "point operations" {
    const p1 = Point{ .x = 3.0, .y = 4.0 };
    const p2 = Point{ .x = 1.0, .y = 2.0 };
    const p3 = p1.add(p2);
    try std.testing.expect(p3.x == 4.0);
}

File: test/data/dart/sample.dart
// Sample Dart file for testing AST parsing
import 'dart:async';
import 'dart:math' as math;

/// A simple Point class with x and y coordinates
class Point {
  final double x;
  final double y;

  /// Creates a new Point with given coordinates
  const Point(this.x, this.y);

  /// Named constructor for origin point
  Point.origin() : x = 0, y = 0;

  /// Factory constructor for creating from map
  factory Point.fromMap(Map<String, double> map) {
    return Point(map['x'] ?? 0, map['y'] ?? 0);
  }

  /// Calculate distance from another point
  double distanceTo(Point other) {
    final dx = x - other.x;
    final dy = y - other.y;
    return math.sqrt(dx * dx + dy * dy);
  }

  /// Add two points
  Point operator +(Point other) => Point(x + other.x, y + other.y);

  @override
  String toString() => 'Point($x, $y)';
}

/// Color enum with RGB values
enum Color {
  red,
  green,
  blue;

  int get rgbValue {
    return switch (this) {
      Color.red => 0xFF0000,
      Color.green => 0x00FF00,
      Color.blue => 0x0000FF,
    };
  }
}

/// Shape abstract class with sealed modifier
sealed class Shape {
  double get area;
}

/// Circle implementation
class Circle extends Shape {
  final double radius;
  Circle(this.radius);

  @override
  double get area => math.pi * radius * radius;
}

/// Rectangle implementation
class Rectangle extends Shape {
  final double width;
  final double height;
  Rectangle(this.width, this.height);

  @override
  double get area => width * height;
}

/// Mixin for printable objects
mixin Printable {
  void printInfo() {
    print(toString());
  }
}

/// Extension on String
extension StringExtension on String {
  String capitalize() {
    if (isEmpty) return this;
    return '${this[0].toUpperCase()}${substring(1)}';
  }
}

/// Generic container class
class Container<T> {
  T? _value;

  T? get value => _value;
  set value(T? newValue) => _value = newValue;

  bool get isEmpty => _value == null;
}

/// Type alias for callback function
typedef StringCallback = void Function(String);

/// Main entry point
Future<void> main() async {
  // Variable declarations
  var message = 'Hello, Dart!';
  final pi = 3.14159;
  const maxSize = 100;
  late String lateVar;

  // Null safety
  String? nullableString;
  String nonNullable = nullableString ?? 'default';

  // Control flow
  if (message.isNotEmpty) {
    print(message);
  } else {
    print('Empty message');
  }

  // For loop
  for (var i = 0; i < 5; i++) {
    if (i == 2) continue;
    if (i == 4) break;
    print(i);
  }

  // For-in loop
  final numbers = [1, 2, 3, 4, 5];
  for (final num in numbers) {
    print(num);
  }

  // While loop
  var counter = 0;
  while (counter < 3) {
    counter++;
  }

  // Do-while loop
  do {
    counter--;
  } while (counter > 0);

  // Switch expression
  final colorName = switch (Color.red) {
    Color.red => 'Red',
    Color.green => 'Green',
    Color.blue => 'Blue',
  };

  // Pattern matching
  final point = Point(3, 4);
  final (x, y) = (point.x, point.y);

  // Try-catch-finally
  try {
    await fetchData();
  } on FormatException catch (e) {
    print('Format error: $e');
  } catch (e, stackTrace) {
    print('Error: $e');
    print(stackTrace);
  } finally {
    print('Cleanup complete');
  }

  // Cascade notation
  final list = <int>[]
    ..add(1)
    ..add(2)
    ..add(3);

  // Collection literals
  final listLiteral = [1, 2, 3];
  final setLiteral = {1, 2, 3};
  final mapLiteral = {'key': 'value', 'count': 42};

  // Spread operator
  final combined = [...listLiteral, 4, 5];

  // Records
  final record = (name: 'Alice', age: 30);
  final (name: userName, age: userAge) = record;

  // Initialize late variable
  lateVar = 'Initialized';
  print(lateVar);
}

/// Async function example
Future<String> fetchData() async {
  await Future.delayed(Duration(seconds: 1));
  return 'Data fetched';
}

/// Stream example
Stream<int> countStream(int max) async* {
  for (var i = 0; i < max; i++) {
    yield i;
  }
}

/// Generator function
Iterable<int> naturals(int n) sync* {
  var k = 0;
  while (k < n) {
    yield k++;
  }
}

/// Private function (starts with underscore)
void _privateHelper() {
  print('This is private');
}

/// Assert example
void validatePositive(int value) {
  assert(value > 0, 'Value must be positive');
}
